ginivision-accounting-network/src/main/java/net/gini/android/vision/accounting/network/GiniVisionAccountingNetworkService.java

```diff
@@ -0,0 +1,559 @@
+                    throws Exception {
+                if (task.isFaulted()) {
+                    final Error error = new Error(getTaskErrorMessage(task));
+                    LOG.error("Document upload failed for {}: {}", document.getId(),
+                            error.getMessage());
+                    callback.failure(error);
+                } else if (task.getResult() != null) {
+                    final net.gini.android.models.Document apiDocument = task.getResult();
+                    LOG.debug("Document upload success for {}: {}", document.getId(),
+                            apiDocument);
+                    mGiniApiDocuments.put(apiDocument.getId(), apiDocument);
+                    callback.success(new Result(apiDocument.getId()));
+                } else {
+                    LOG.debug("Document upload cancelled for {}", document.getId());
+                    callback.cancelled();
+                }
+                return null;
+            }
+        }, Task.UI_THREAD_EXECUTOR);
+        return new NoOpCancellationToken();
+    }
+
+    private String getTaskErrorMessage(@NonNull final Task task) {
+        if (!task.isFaulted()) {
+            return "";
+        }
+        final String errorMessage = task.getError().getMessage();
+        return errorMessage != null ? errorMessage : task.getError().toString();
+    }
+
+    @Override
+    public CancellationToken delete(@NonNull final String giniApiDocumentId,
+            @NonNull final GiniVisionNetworkCallback<Result, Error> callback) {
+        LOG.debug("Delete document with api id {}", giniApiDocumentId);
+        mGiniApi.getDocumentTaskManager().deleteDocument(giniApiDocumentId)
+                .continueWith(new Continuation<String, Void>() {
+                    @Override
+                    public Void then(final Task<String> task) throws Exception {
+                        if (task.isFaulted()) {
+                            final Error error = new Error(getTaskErrorMessage(task));
+                            LOG.error("Document deletion failed for api id {}: {}",
+                                    giniApiDocumentId,
+                                    error.getMessage());
+                            callback.failure(error);
+                        } else if (task.getResult() != null) {
+                            LOG.debug("Document deletion success for api id {}", giniApiDocumentId);
+                            callback.success(new Result(giniApiDocumentId));
+                        } else {
+                            LOG.debug("Document deletion cancelled for api id {}",
+                                    giniApiDocumentId);
+                            callback.cancelled();
+                        }
+                        return null;
+                    }
+                }, Task.UI_THREAD_EXECUTOR);
+        return new NoOpCancellationToken();
+    }
+
+    @SuppressWarnings("PMD.ExcessiveMethodLength")
```

Comment:
Seeing this I guess it's impossible to reduce this method length, isn't it?

Reply:
Well, given enough time and money, it's doable I guess 😛.

Reply:
Fair point 👍